src/main/java/io/cryostat/recordings/JvmIdHelper.java

```diff
@@ -38,120 +38,107 @@
+        this.connectionTimeoutSeconds = connectionTimeoutSeconds;
         this.logger = logger;
-        this.jvmIdMap = new ConcurrentHashMap<>();
-    }
-
-    protected String computeJvmId(ConnectionDescriptor cd) {
-        String targetId = cd.getTargetId();
-        if (targetId.equals(RecordingArchiveHelper.ARCHIVES)
-                || targetId.equals(RecordingArchiveHelper.UPLOADED_RECORDINGS_SUBDIRECTORY)) {
-            return RecordingArchiveHelper.UPLOADED_RECORDINGS_SUBDIRECTORY;
-        }
-        try {
-            if (cd.getCredentials().isEmpty()) {
-                cd =
-                        new ConnectionDescriptor(
-                                targetId, credentialsManager.getCredentialsByTargetId(targetId));
-            }
-            final ConnectionDescriptor desc = cd;
-            return this.targetConnectionManager.executeConnectedTask(
-                    desc,
-                    connection -> {
-                        try {
-                            return connection.getJvmId();
-                        } catch (Exception e) {
-                            if (e.getCause() instanceof SecurityException) {
-                                // don't have credentials to access target
-                                logger.warn(
-                                        "Target {} credentials are invalid", desc.getTargetId());
-                            } else {
-                                logger.warn(e);
-                            }
-                            return null;
-                        }
-                    });
-        } catch (Exception e) {
-            logger.warn(e);
-            return null;
-        }
+        this.ids =
+                Caffeine.newBuilder()
+                        .executor(executor)
+                        .scheduler(scheduler)
+                        .<String, String>removalListener(
+                                (targetId, jvmId, cause) -> emit(IdEvent.INVALIDATED, jvmId))
+                        .buildAsync(new IdLoader());
+
+        platform.addTargetDiscoveryListener(
+                tde -> {
+                    switch (tde.getEventKind()) {
+                        case LOST:
+                            String targetId = tde.getServiceRef().getServiceUri().toString();
+                            ids.synchronous().invalidate(targetId);
+                            break;
+                        default:
+                            // ignored
+                            break;
+                    }
+                });
     }
 
-    public String getJvmId(ConnectionDescriptor connectionDescriptor) throws JvmIdGetException {
-        String targetId = connectionDescriptor.getTargetId();
+    private CompletableFuture<String> computeJvmId(String targetId) throws ScriptException {
+        CompletableFuture<String> future;
         // FIXME: this should be refactored after the 2.2.0 release
         if (targetId == null
```

Comment:
One small thing, if a targetId == "uploads", a log will still show something like this 
```
INFO: JVM ID: uploads -> uploads
```
Doesn't really matter but yeah. :-)